### PR TITLE
Change SizeLimit to a pointer

### LINKED ISSUE #50121: SizeLimit violates k8s API convention
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**: 
https://github.com/kubernetes/community/pull/306 added field `SizeLimit` in PR https://github.com/kubernetes/kubernetes/commit/85f030c2aad00c54d498646e4151e7fad930c7c3. The field is marked as _optional_ yet has the non-pointer type `resource.Quantity`. This is against the API convention that requires optional fields to be pointers. 

As a consequence, k8s client-go library does not omit this field in serialization, breaking k8s 1.6 compatibility for this API.

**What you expected to happen**:
client-go 1.7 works with k8s 1.6 in pass-through object processing pipeline.

**How to reproduce it (as minimally and precisely as possible)**:
See https://github.com/kubernetes/client-go/issues/257#issuecomment-319552550

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.7
- Cloud provider or hardware configuration**: GKE
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

cc @caesarxuchao

### PR DESCRIPTION
**Pull request by jingxu97**
This PR fixes #50121

```release-note
The `emptyDir.sizeLimit` field is now correctly omitted from API requests and responses when unset.
```

### CODE REVIEW

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -383,8 +383,7 @@ func validateVolumeSource(source *api.VolumeSource, fldPath *field.Path) field.E
 	if source.EmptyDir != nil {
 		numVolumes++
 		if !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
-			unsetSizeLimit := resource.Quantity{}
-			if unsetSizeLimit.Cmp(source.EmptyDir.SizeLimit) != 0 {
+			if source.EmptyDir.SizeLimit != nil {
```
Comment (liggitt) [inline pkg/api/validation/validation.go]: 1.7 clients will send "0" and we could have 1.7 data stored as 0... need to tolerate that

Comment (jingxu97) [inline pkg/api/validation/validation.go]: I added a check. Will this work for the case of data stored as 0? Thanks!

**Comment by jingxu97** [comment]
@liggitt I updated the PR, PTAL. Thanks!

**Comment by jingxu97** [comment]
@liggitt I addressed your comment. Could you please help review it again? Thanks!

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (sttts) [inline pkg/api/validation/validation.go]: Can you add a unit test for that?

**Comment by kyessenov** [comment]
Is there a date when this gets merged?

**Comment by liggitt** [comment]
one question on the zero compare (and associated validation), needs rebase, then LGTM

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -383,8 +383,7 @@ func validateVolumeSource(source *api.VolumeSource, fldPath *field.Path) field.E
 	if source.EmptyDir != nil {
 		numVolumes++
 		if !utilfeature.DefaultFeatureGate.Enabled(features.LocalStorageCapacityIsolation) {
-			unsetSizeLimit := resource.Quantity{}
-			if unsetSizeLimit.Cmp(source.EmptyDir.SizeLimit) != 0 {
+			if source.EmptyDir.SizeLimit != nil && source.EmptyDir.SizeLimit.Cmp(resource.Quantity{}) > 0 {
```
Comment (jingxu97) [inline pkg/api/validation/validation.go]: I am thinking I only need to check sizeLimit != nil. No matter >0, <0 or =0 should not be allowed

Comment (liggitt) [inline pkg/api/validation/validation.go]: 1.7 clients (and existing data) could have `0`, which you now have to tolerate.

**Comment by liggitt** [comment]
/lgtm
let's gather approvals, then do a final rebase

**Comment by jingxu97** [comment]
cc @vishh PTAL, thanks!

**Comment by jingxu97** [comment]
cc @davidopp Could you please help approve this PR? Thanks!

**Review by timothysc — COMMENTED** [review]
/approve

**Comment by liggitt** [comment]
`
W0829 00:13:59.563] 2017/08/29 00:13:59 error running compiler: exit status 1
W0829 00:13:59.563] Use --strategy=GoCompile=standalone to disable sandboxing for the failing actions.
I0829 00:13:59.664] test/e2e_node/local_storage_isolation_eviction_test.go:221: cannot use *resource.NewQuantity(int64(100000), resource.BinarySI) (type resource.Quantity) as type *resource.Quantity in field value
`

**Comment by jingxu97** [comment]
@liggitt I already changed the code in test/e2e_node/local_storage_isolation_eviction_test.go:221:  to use pointer, could not figure out why it has the compile error..

File: staging/src/k8s.io/api/core/v1/types.go [hunk 3]
```diff
@@ -739,7 +739,7 @@ type EmptyDirVolumeSource struct {
 	// The default is nil which means that the limit is undefined.
 	// More info: http://kubernetes.io/docs/user-guide/volumes#emptydir
 	// +optional
-	SizeLimit resource.Quantity `json:"sizeLimit,omitempty" protobuf:"bytes,2,opt,name=sizeLimit"`
+	SizeLimit *resource.Quantity `json:"sizeLimit,omitempty" protobuf:"bytes,2,opt,name=sizeLimit"`
```
Comment (janetkuo) [inline staging/src/k8s.io/api/core/v1/types.go]: Doesn't this break backward compatibility?

Comment (liggitt) [reply staging/src/k8s.io/api/core/v1/types.go]: Actually, the introduction of the non-pointer field broke compatibility, and this restored it

**Pull request merged**